python/ray/data/_internal/lazy_block_list.py

```diff
@@ -162,7 +162,11 @@ def split(self, split_size: int) -> List["LazyBlockList"]:
         block_partition_meta_refs = np.array_split(
             self._block_partition_meta_refs, num_splits
         )
-        cached_metadata = np.array_split(self._cached_metadata, num_splits)
+        k, m = divmod(len(self._cached_metadata), num_splits)
+        cached_metadata = [
+            self._cached_metadata[i * k + min(i, m) : (i + 1) * k + min(i + 1, m)]
+            for i in range(len(self._cached_metadata))
+        ]
```

Comment:
Hmm it seems like this `cached_metadata` should always be homogeneous, i.e. it should always contain a `BlockMetadata` for each element. Do you know how this heterogeneity is happening?

Reply:
We set only the first block when figuring out the schema, so e.g. with this:
```
import ray

inputs = ["example://iris.csv"] * 100
ds = ray.data.read_csv(inputs, parallelism=10)
print("before:", ds._plan._in_blocks._cached_metadata)
ds.schema()
print("after:", ds._plan._in_blocks._cached_metadata)
ds._plan._in_blocks.split(2)
```
It's producing:
```
before: [None, None, None, None, None, None, None, None, None, None]
after: [[BlockMetadata(num_rows=1500, size_bytes=66500, schema=sepal.length: double
sepal.width: double
petal.length: double
petal.width: double
variety: string, input_files=array(['/home/ubuntu/ray/python/ray/data/examples/data/iris.csv',
       '/home/ubuntu/ray/python/ray/data/examples/data/iris.csv',
       '/home/ubuntu/ray/python/ray/data/examples/data/iris.csv',
       '/home/ubuntu/ray/python/ray/data/examples/data/iris.csv',
       '/home/ubuntu/ray/python/ray/data/examples/data/iris.csv',
       '/home/ubuntu/ray/python/ray/data/examples/data/iris.csv',
       '/home/ubuntu/ray/python/ray/data/examples/data/iris.csv',
       '/home/ubuntu/ray/python/ray/data/examples/data/iris.csv',
       '/home/ubuntu/ray/python/ray/data/examples/data/iris.csv',
       '/home/ubuntu/ray/python/ray/data/examples/data/iris.csv'],
      dtype='<U55'), exec_stats={'wall_time_s': 0.36291642487049103, 'cpu_time_s': 0.3410033880000001, 'node_id': 'f3e389087180baf4bcde82efe3873d1139be957718e69465786af17d'})], None, None, None, None, None, None, None, None, None]
```

Reply:
Could we pull this out into an array split util and use it for the above splits too?

Otherwise it's not clear this is doing the same thing as array split.

Reply:
Sg, done.